File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/compare/ProductCompareActivity.java
package openfoodfacts.github.scrachx.openfood.features.compare;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Toast;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.ArrayList;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.databinding.ActivityProductComparisonBinding;
import openfoodfacts.github.scrachx.openfood.features.listeners.CommonBottomListenerInstaller;
import openfoodfacts.github.scrachx.openfood.features.scan.ContinuousScanActivity;
import openfoodfacts.github.scrachx.openfood.features.shared.BaseActivity;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.utils.PhotoReceiverHandler;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

public class ProductCompareActivity extends BaseActivity {
    private ActivityProductComparisonBinding binding;
    private PhotoReceiverHandler photoReceiverHandler;
    private RecyclerView.Adapter<ProductCompareAdapter.ProductComparisonViewHolder> productComparisonAdapter;
    private ArrayList<Product> products = new ArrayList<>();

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
    }

    public static void start(Context context) {
        Intent starter = new Intent(context, ProductCompareActivity.class);
        context.startActivity(starter);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityProductComparisonBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setTitle(getString(R.string.compare_products));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        photoReceiverHandler = new PhotoReceiverHandler(newPhotoFile -> ((ProductCompareAdapter) productComparisonAdapter).setImageOnPhotoReturn(newPhotoFile));

        if (getIntent().getExtras() != null && getIntent().getBooleanExtra("product_found", false)) {
            products = (ArrayList<Product>) getIntent().getExtras().getSerializable("products_to_compare");
            if (getIntent().getBooleanExtra("product_already_exists", false)) {
                Toast.makeText(this, getString(R.string.product_already_exists_in_comparison), Toast.LENGTH_SHORT).show();
            }
        }

        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        binding.productComparisonRv.setLayoutManager(layoutManager);

        productComparisonAdapter = new ProductCompareAdapter(products, this);
        binding.productComparisonRv.setAdapter(productComparisonAdapter);

        binding.productComparisonButton.setOnClickListener(v -> {
            if (Utils.isHardwareCameraInstalled(ProductCompareActivity.this)) {
                if (ContextCompat.checkSelfPermission(ProductCompareActivity.this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                    if (ActivityCompat.shouldShowRequestPermissionRationale(ProductCompareActivity.this, Manifest.permission.CAMERA)) {
                        new MaterialDialog.Builder(ProductCompareActivity.this)
                            .title(R.string.action_about)
                            .content(R.string.permission_camera)
                            .neutralText(R.string.txtOk)
                            .onNeutral((dialog, which) -> ActivityCompat
                                .requestPermissions(ProductCompareActivity.this, new String[]{Manifest.permission.CAMERA}, Utils.MY_PERMISSIONS_REQUEST_CAMERA))
                            .show();
                    } else {
                        ActivityCompat.requestPermissions(ProductCompareActivity.this, new String[]{Manifest.permission.CAMERA}, Utils.MY_PERMISSIONS_REQUEST_CAMERA);
                    }
                } else {
                    Intent intent = new Intent(ProductCompareActivity.this, ContinuousScanActivity.class);
                    intent.putExtra("compare_product", true);
                    intent.putExtra("products_to_compare", products);
                    startActivity(intent);
                }
            }
        });

        CommonBottomListenerInstaller.install(this, binding.navigationBottomInclude.bottomNavigation);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        photoReceiverHandler.onActivityResult(this, requestCode, resultCode, data);
    }

    @Override
    public void onResume() {
        super.onResume();
        CommonBottomListenerInstaller.selectNavigationItem(binding.navigationBottomInclude.bottomNavigation, R.id.compare_products);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/compare/ProductCompareActivity.kt
package openfoodfacts.github.scrachx.openfood.features.compare

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import androidx.activity.result.launch
import androidx.activity.viewModels
import androidx.core.view.isVisible
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.ActivityProductComparisonBinding
import openfoodfacts.github.scrachx.openfood.features.compare.ProductCompareViewModel.SideEffect
import openfoodfacts.github.scrachx.openfood.features.product.view.ProductViewActivityStarter
import openfoodfacts.github.scrachx.openfood.features.shared.BaseActivity
import openfoodfacts.github.scrachx.openfood.features.simplescan.SimpleScanActivityContract
import openfoodfacts.github.scrachx.openfood.images.ProductImage
import openfoodfacts.github.scrachx.openfood.listeners.CommonBottomListenerInstaller.installBottomNavigation
import openfoodfacts.github.scrachx.openfood.listeners.CommonBottomListenerInstaller.selectNavigationItem
import openfoodfacts.github.scrachx.openfood.models.Product
import openfoodfacts.github.scrachx.openfood.models.ProductImageField
import openfoodfacts.github.scrachx.openfood.repositories.ProductRepository
import openfoodfacts.github.scrachx.openfood.utils.Intent
import openfoodfacts.github.scrachx.openfood.utils.PhotoReceiverHandler
import javax.inject.Inject

@AndroidEntryPoint
class ProductCompareActivity : BaseActivity() {

    private lateinit var binding: ActivityProductComparisonBinding

    private val viewModel: ProductCompareViewModel by viewModels()

    @Inject
    lateinit var client: ProductRepository

    @Inject
    lateinit var picasso: Picasso

    @Inject
    lateinit var sharedPreferences: SharedPreferences

    @Inject
    lateinit var productViewActivityStarter: ProductViewActivityStarter

    @Inject
    lateinit var photoReceiverHandler: PhotoReceiverHandler

    private val scanProductContract = registerForActivityResult(SimpleScanActivityContract()) { barcode ->
        barcode?.let { viewModel.barcodeDetected(it) }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityProductComparisonBinding.inflate(layoutInflater)
        setContentView(binding.root)

        title = getString(R.string.compare_products)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.navigationBottomInclude.bottomNavigation.installBottomNavigation(this)

        (intent.extras?.getSerializable(KEY_PRODUCTS_TO_COMPARE) as? Product)?.let {
            viewModel.addProductToCompare(it)
        }

        lifecycleScope.launch {
            launch {
                viewModel.sideEffectFlow
                    .flowWithLifecycle(lifecycle)
                    .collect {
                        when (it) {
                            is SideEffect.ProductAlreadyAdded -> showProductAlreadyAddedDialog()
                            is SideEffect.ProductNotFound -> showProductNotFoundDialog()
                            is SideEffect.ConnectionError -> showConnectionErrorDialog()
                        }
                    }
            }
            launch {
                viewModel.productsFlow
                    .flowWithLifecycle(lifecycle)
                    .distinctUntilChanged()
                    .collect { createAdapter(it) }
            }
            launch {
                viewModel.loadingVisibleFlow
                    .flowWithLifecycle(lifecycle)
                    .distinctUntilChanged()
                    .collect {
                        binding.comparisonProgressView.isVisible = it
                    }
            }
        }


        binding.productComparisonButton.setOnClickListener { openScanActivity() }
    }

    override fun startScanActivity() {
        scanProductContract.launch()
    }


    public override fun onResume() {
        super.onResume()
        binding.navigationBottomInclude.bottomNavigation.selectNavigationItem(R.id.compare_products)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        photoReceiverHandler.onActivityResult(this, requestCode, resultCode, data) {
            (binding.productComparisonRv.adapter as ProductCompareAdapter).onImageReturned(it)
        }
    }

    private fun createAdapter(products: List<ProductCompareViewModel.CompareProduct>) {
        // Create adapter
        val productComparisonAdapter = ProductCompareAdapter(
            products,
            this,
            this,
            client,
            picasso,
            viewModel.getCurrentLanguage()
        ).apply {
            imageReturnedListener = { product, file ->

                val image = ProductImage(
                    product.code,
                    ProductImageField.FRONT,
                    file,
                    viewModel.getCurrentLanguage()
                ).apply { filePath = file.absolutePath }

                lifecycleScope.launch { client.postImg(image) }
                product.imageUrl = file.absolutePath
            }

            fullProductClickListener = {
                val barcode = it.code
                productViewActivityStarter.openProduct(barcode, this@ProductCompareActivity)
            }
        }

        // Update product comparison recyclerview
        binding.productComparisonRv.layoutManager = LinearLayoutManager(this@ProductCompareActivity, HORIZONTAL, false)
        binding.productComparisonRv.adapter = productComparisonAdapter

    }

    private fun showProductAlreadyAddedDialog() {
        MaterialAlertDialogBuilder(this)
            .setMessage(R.string.product_already_exists_in_comparison)
            .setPositiveButton(R.string.ok_button) { dialog, _ ->
                dialog.dismiss()
            }
            .show()
    }

    private fun showProductNotFoundDialog() {
        MaterialAlertDialogBuilder(this)
            .setMessage(R.string.product_does_not_exist_please_add_it)
            .setPositiveButton(R.string.ok_button) { dialog, _ ->
                dialog.dismiss()
            }
            .show()
    }

    private fun showConnectionErrorDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.alert_dialog_warning_title)
            .setMessage(R.string.txtConnectionError)
            .setPositiveButton(R.string.ok_button) { dialog, _ ->
                dialog.dismiss()
            }
            .show()
    }

    companion object {
        const val KEY_PRODUCTS_TO_COMPARE = "products_to_compare"

        @JvmStatic
        fun start(context: Context, product: Product? = null) {
            context.startActivity(Intent<ProductCompareActivity>(context) {
                if (product != null) putExtra(KEY_PRODUCTS_TO_COMPARE, product)
            })
        }
    }
}
